cmd/streaming-v4-unsigned.go

```diff
@@ -0,0 +1,257 @@
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+)
+
+// newUnsignedV4ChunkedReader returns a new s3UnsignedChunkedReader that translates the data read from r
+// out of HTTP "chunked" format before returning it.
+// The s3ChunkedReader returns io.EOF when the final 0-length chunk is read.
+func newUnsignedV4ChunkedReader(req *http.Request, trailer bool) (io.ReadCloser, APIErrorCode) {
+	if trailer {
+		// Discard anything unsigned.
+		req.Trailer = make(http.Header)
+		trailers := req.Header.Values(awsTrailerHeader)
+		for _, key := range trailers {
+			req.Trailer.Add(key, "")
+		}
+	} else {
+		req.Trailer = nil
+	}
+	return &s3UnsignedChunkedReader{
+		trailers: req.Trailer,
+		reader:   bufio.NewReader(req.Body),
+		buffer:   make([]byte, 64*1024),
+	}, ErrNone
+}
+
+// Represents the overall state that is required for decoding a
+// AWS Signature V4 chunked reader.
+type s3UnsignedChunkedReader struct {
+	reader   *bufio.Reader
+	trailers http.Header
+
+	buffer []byte
+	offset int
+	err    error
+	debug  bool
+}
+
+func (cr *s3UnsignedChunkedReader) Close() (err error) {
+	return cr.err
+}
+
+// Read - implements `io.Reader`, which transparently decodes
+// the incoming AWS Signature V4 streaming signature.
+func (cr *s3UnsignedChunkedReader) Read(buf []byte) (n int, err error) {
+	// First, if there is any unread data, copy it to the client
+	// provided buffer.
+	if cr.offset > 0 {
+		n = copy(buf, cr.buffer[cr.offset:])
+		if n == len(buf) {
+			cr.offset += n
+			return n, nil
+		}
+		cr.offset = 0
+		buf = buf[n:]
+	}
+	// mustRead reads from input and compares against provided slice.
+	mustRead := func(b ...byte) error {
+		for _, want := range b {
+			got, err := cr.reader.ReadByte()
+			if err == io.EOF {
+				return io.ErrUnexpectedEOF
+			}
+			if got != want {
+				if cr.debug {
+					fmt.Printf("mustread: want: %q got: %q\n", string(want), string(got))
+				}
+				return errMalformedEncoding
+			}
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	var size int
+	for {
+		b, err := cr.reader.ReadByte()
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+		if err != nil {
+			cr.err = err
+			return n, cr.err
+		}
+		if b == '\r' { // \r\n denotes end of size.
+			err := mustRead('\n')
+			if err != nil {
+				cr.err = err
+				return n, cr.err
+			}
+			break
+		}
+
+		// Manually deserialize the size since AWS specified
+		// the chunk size to be of variable width. In particular,
+		// a size of 16 is encoded as `10` while a size of 64 KB
+		// is `10000`.
```

Comment:
And actually I tried to use MinIO with `STREAMING-UNSIGNED-PAYLOAD-TRAILER` but failed, so I looked around and wrote the following patch - https://github.com/minio/minio/commit/7d4c0c311c06d296652a035512dd7b22e352d10f

As far as I understand, in case of `STREAMING-UNSIGNED-PAYLOAD-TRAILER`, there are no "aws specified chunks" only regular plain HTTP chunks, so all of this looks superfluous.

Although I'm not sure about checksum part in the patch, apparently the aws-cpp-sdk simply does not support `STREAMING-UNSIGNED-PAYLOAD-TRAILER` properly, since when I tried to send request to the AWS S3, it was complaining:

```
S3Exception name IncompleteBody, Message: Unable to parse ExceptionName: IncompleteBody Message: The request body terminated unexpectedly
```

@klauspost what do you think?

Reply:
@azat 
Our [minio-go](https://github.com/minio/minio-go) implementation is confirmed to work against AWS:

```
PUT /1z5dsknvswcxo92b6sdfajulvwfxh2?partNumber=2&uploadId=mj6EhLRx2c1plArOgolRk3hgCtFQP6408LqkU_JR910Qu.QvFUTJGAwqAUddj3IglD51NgCVNLncjd2WLG6XlkXmFixKz9M1.ifxJagF0pR2d.qQWnEqD_m3M3DkreQW HTTP/1.1
Host: minio-go-test-nu40trprcadk2wmv.s3.dualstack.us-east-1.amazonaws.com
User-Agent: MinIO (windows; amd64) minio-go/v7.0.60 MinIO-go-FunctionalTest/0.1.0
Content-Length: 4889394
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20231005/us-east-1/s3/aws4_request, SignedHeaders=content-encoding;host;x-amz-content-sha256;x-amz-date;x-amz-decoded-content-length;x-amz-trailer, Signature=**REDACTED**
Content-Encoding: aws-chunked
X-Amz-Content-Sha256: STREAMING-UNSIGNED-PAYLOAD-TRAILER
X-Amz-Date: 20231005T170945Z
X-Amz-Decoded-Content-Length: 4888682
X-Amz-Trailer: x-amz-checksum-crc32c
Accept-Encoding: gzip

HTTP/1.1 200 OK
Content-Length: 0
Date: Thu, 05 Oct 2023 17:09:46 GMT
Etag: "2e17afc721dce2bca8572e594fa4451d"
Server: AmazonS3
X-Amz-Checksum-Crc32c: C/etzg==
X-Amz-Id-2: qYB5u7L3L1T9w18BoUTwYmkhF05TdP6IrVPQxBnvTSp1JiVbaWzQK8Jup7EE2mINcP5CcRFm/DU=
X-Amz-Request-Id: B70CB93P78ZVZBV5
X-Amz-Server-Side-Encryption: AES256
```

Reply:
@azat Since the CPP library fails on AWS, I suspect it doesn't support it.

Reply:
>Our [minio-go](https://github.com/minio/minio-go) implementation is confirmed to work against AWS:

It seems that there is no cli that I can use?
(I'm not a go developer actually, just a needy user)

But minio-go is not the same as minio, right?
Or you are saying that it implements protocol in the same fashion as this implementation in the server, and it send "aws chunks" even for unsigned requests? And it works with AWS S3 and MinIO?

When I tested this `STREAMING-UNSIGNED-PAYLOAD-TRAILER`, using C++ sdk, I got this:
- S3: `Message: Unable to parse ExceptionName: IncompleteBody Message: The request body terminated unexpectedly,`
- GCS: works
- R2 (cloudflare): `Message: Unable to parse ExceptionName: NotImplemented Message: STREAMING-UNSIGNED-PAYLOAD-TRAILER not implemented,`

So maybe the format was wrong indeed, but this means that GCS implemented it not correctly.

Reply:
minio-go implementation works for both MinIO and AWS S3 - that's what @klauspost meant. 

Meaning our implementation is correct and to the spec.

Reply:
@azat TBH I referenced the documentation and the AWS Java SDK IIRC.

Considering their SDK doesn't work against their own servers, I don't really have much confidence in it. I suggest you discuss that with AWS.

As Harsha mentioned our client implementation works against S3 (and it was pretty strict in what it accepted). It seems reasonable to assume that the matching server implementation is correct as well. I don't really care what GCS does.